### DISCUSSION TITLE
Redirect www to non-www in NextJS 12

### DISCUSSION THREAD

**Discussion by caleb531** [vercel/next.js#33554] [Help] [answered]
Hi,

I have a React web app (under a custom domain) that I am converting to NextJS (which so far is _awesome_). I want to 301-redirect my app (all pages, preferably) from www to non-www. Before migrating to NextJS, I wrote some Express middleware to take care of this for me:

```ts
// This is all TypeScript, BTW
import { NextFunction, Request, Response } from 'express';

// A regular expression used for matching a www domain
const wwwRegex = /^www\./;

// Redirect www requests to non-www
export function redirectWwwToNonWww(req: Request, res: Response, next: NextFunction): void {
  if (wwwRegex.test(req.headers.host)) {
    const newHost = req.headers.host.replace(wwwRegex, '');
    return res.redirect(301, `${req.protocol}://${newHost}${req.originalUrl}`);
  }
  next();
}
```

I am trying to accomplish similar behavior in NextJS. After scouring the internet and reading up on NextJS features like `_middleware.ts` and `get*Props` functions, I've landed here:

```ts
import http from 'http';
import { GetServerSideProps } from 'next';
import App from '../components/app/App';

// A regular expression used for matching a www domain
const wwwRegex = /^www\./;

function redirectWwwToNonWww(req: http.IncomingMessage, res: http.ServerResponse) {
  if (wwwRegex.test(String(req.headers.host))) {
    const newHost = String(req.headers.host).replace(wwwRegex, '');
    // Redirect to plain http for now (while testing on www.localhost)
    res.writeHead(301, { Location: `http://${newHost}` });
    res.end();
  }
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  redirectWwwToNonWww(context.req, context.res);
  return { props: {} };
};

export default App;
```

The problem is, when I visit `http://www.localhost:3000` the page just hangs forever, when I expect it to redirect to `http://localhost:3000`. _**Is there a way to redirect www to non-www without writing a custom server?**_ (I'd prefer to keep features like serverless functions, etc.)

If it helps, here's a link to [my specific commit and code repository.](https://github.com/caleb531/faith-dashboard/commit/8075aa2339fc073fd7da783ce9fb78812a1daaf0)

Thank you so much in advance! And really digging NextJS otherwise.
Caleb

*P.S. I have tried using `_middleware.ts`, but it doesn't seem to expose the request hostname.*

**Comment by caleb531** [comment]
For anyone having the same problem, I think I figured it out. The key is not to use `getServerSideProps`, but rather, use `getInitialProps` in `pages/_app.tsx`.

```ts
import http from 'http';
import App, { AppContext, AppProps } from 'next/app';

function AppWrapper({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

// A regular expression used for matching a www domain
const wwwRegex = /^www\./;

function redirectWwwToNonWww(req: http.IncomingMessage, res: http.ServerResponse) {
  if (wwwRegex.test(String(req.headers.host))) {
    const newHost = String(req.headers.host).replace(wwwRegex, '');
    res.writeHead(301, { location: `http://${newHost}` });
    res.end();
  }
}

// Run server-side code on each request
AppWrapper.getInitialProps = async (appContext: AppContext) => {
  redirectWwwToNonWww(appContext.ctx.req, appContext.ctx.res);
  const appProps = await App.getInitialProps(appContext);
  return { ...appProps };
};

export default AppWrapper;
```

There _is_ one caveat mentioned by the docs on Custom App:

> This disables the ability to perform automatic static optimization, causing every page in your app to be server-side rendered.

But this makes sense, because each page needs to hit the redirect code before loading. Anyway, I hope this helps someone!

**Update:** It seems that adding `getInitialProps` to `_app.tsx` completely breaks NextJS's [Font Optimization](https://nextjs.org/docs/basic-features/font-optimization), so if you are using Google Fonts, using `getInitialProps` will prevent your custom fonts from loading.

**Comment by caleb531** [comment, marked as answer]
Just an update for anyone searching about this: I found a better solution that allows for the preservation of static pages and Font Optimization while still enabling the redirect behavior:

It took some fiddling, but you can create a `_middleware.tsx` file within the `pages/` directory with the following code:

```tsx
import { NextRequest, NextResponse } from 'next/server';

export function middleware(req: NextRequest) {
  const host = req.headers.get('host');
  const wwwRegex = /^www\./;
  // This redirect will only take effect on a production website (on a non-localhost domain)
  if (wwwRegex.test(host) && !req.headers.get('host').includes('localhost')) {
    const newHost = host.replace(wwwRegex, '');
    return NextResponse.redirect(`https://${newHost}${req.nextUrl.pathname}`, 301);
  }
  return NextResponse.next();
}
```

[Link to Documentation on NextJS Middleware](https://nextjs.org/docs/middleware)

**Reply by devolasvegas** [reply]
Hi, @caleb531. I wonder if, perhaps, checking out the [Next.Js redirects](https://nextjs.org/docs/api-reference/next.config.js/redirects) and [rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites) in their docs might have a more elegant solution for you than having to write a middleware?

**Reply by caleb531** [reply]
@devolasvegas I recall trying that initially with no luck. Besides, all of the examples in the docs you referenced treat `source` as starting from the root of the domain (without specifying the actual domain name). And even if you could specify the domain name as part of the `source` value, it probably wouldn't be domain-agnostic.

Regardless, the middleware solution has been working well for me, and in fact may not even be necessary in my case since Vercel's servers seem to perform the redirect automatically (I currently host all my NextJS projects on Vercel).

**Reply by devolasvegas** [reply]
I'm on Vercel too and set up a 301 for `www` right away. Didn't even consider it might already automatically redirect. I may have to try that out.

Just for future reference, and in case any body else stumbles across this, I did get it working with rewrites in `next.config.js`. I believe it was something like:

```javascript
{
        source: '/:path*',
        has: [
          {
            type: 'host',
            value: 'www',
          },
        ],
        destination: '/:path*'
},
```

But then I ran into CORS issues with trying to call functions on the apex domain. Easy enough to fix, but I didn't feel like it at the time. 😄

**Reply by caleb531** [reply]
@devolasvegas Nice! I suppose you might also want to add `permanent: true` in there so that you can enforce a canonical domain (from an SEO perspective). If I read [the docs](https://nextjs.org/docs/api-reference/next.config.js/redirects) correctly, should be something like:

```js
{
        source: '/:path*',
        has: [
          {
            type: 'host',
            value: 'www',
          },
        ],
        destination: '/:path*',
        permanent: true
},
```

I will add that the nice thing about having the redirects be configuration-based (rather than relying on whatever Vercel's behavior is) is that your project becomes a little more provider-agnostic.

**Reply by davidsiemers** [reply]
Heads up for folks potentially implementing this solution, the middleware redirect laid out here doesn't pass the request's search parameters. If that's wanted, it should look more like this:
```
return NextResponse.redirect(
  `${host}/${req.nextUrl.pathname}?${req.nextUrl.searchParams}`,
  308
)
````

Also, note that I'm using the `308` HTTP status code instead of `301` to preserve the HTTP method--that's currently how most frameworks are redirecting, including [Next](https://nextjs.org/docs/api-reference/next.config.js/redirects )

**Reply by glauber-cruz** [reply]
> Just an update for anyone searching about this: I found a better solution that allows for the preservation of static pages and Font Optimization while still enabling the redirect behavior:
> 
> It took some fiddling, but you can create a `_middleware.tsx` file within the `pages/` directory with the following code:
> 
> ```tsx
> import { NextRequest, NextResponse } from 'next/server';
> 
> export function middleware(req: NextRequest) {
>   const host = req.headers.get('host');
>   const wwwRegex = /^www\./;
>   // This redirect will only take effect on a production website (on a non-localhost domain)
>   if (wwwRegex.test(host) && !req.headers.get('host').includes('localhost')) {
>     const newHost = host.replace(wwwRegex, '');
>     return NextResponse.redirect(`https://${newHost}${req.nextUrl.pathname}`, 301);
>   }
>   return NextResponse.next();
> }
> ```
> 
> [Link to Documentation on NextJS Middleware](https://nextjs.org/docs/middleware)

Thanks dude, u help me a lot in my job S2

**Reply by SlavaMelnyk** [reply]
Thanks for your investigations
Can't we reuse 'host' value in a place of '!req.headers.get('host').includes('localhost')' ?

**Reply by mc-stephen** [reply]
Thanks for this, this has been bugging me a lot and this did fixed the issue for me.

**Comment by zolero** [comment]
For people that are still struggling with this issue. We had nextjs behind a nginx proxy server. (**jwilder**). We needed to check if **x-forwarded-proto** was set.
 >Also we needed to put middleware.ts in src/middleware.ts, not in src/pages/middleware.ts.

```ts
import { NextRequest, NextResponse } from "next/server";

export function middleware(req: NextRequest) {
    const host = req.headers.get("host");
    if (host?.slice(0, 4) === "www.") {
        return NextResponse.redirect(
            (req.headers.get("x-forwarded-proto") || req.nextUrl.protocol.slice(0,-1)) +
                "://" +
                host.slice(4) +
                req.nextUrl.pathname,
            301
        );
    }
    return NextResponse.next();
}
```

And we had to slice the first character of original NextURL protocol. Because it returns: https: instead of https. Weird behaviour.
```ts
req.nextUrl.protocol.slice(0,-1)
```

**Reply by atomoc** [reply]
> Also we needed to put middleware.ts in src/middleware.ts, not in src/pages/middleware.ts.

very strange, but I needed to place `/middleware.ts` in the root of the project
